specs/rollup-node.md

```diff
@@ -1,167 +1,227 @@
+
+- `timestamp` is set to the timestamp of the L1 block.
 - `random` is set to the *random* L1 block attribute
 - `suggestedFeeRecipient` is set to the zero-address for deposit-blocks, since there is no sequencer.
-- `transactions` is an array of the derived deposits, all encoded in the [EIP-2718] format.
+- `transactions` is an array of the derived deposits, encoded as per the two preceding sections.
 
-[expanded version of PayloadAttributesV1]: exec-engine.md#extended-payloadattributesv1
-[unix time]: https://en.wikipedia.org/wiki/Unix_time
-[EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718
-[EIP-2930]: https://eips.ethereum.org/EIPS/eip-2930
+[expanded-payload]: exec-engine.md#extended-payloadattributesv1
+[`PayloadAttributesV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1
 
-### Output derivation
+------------------------------------------------------------------------------------------------------------------------
 
-Building a full block requires the earlier [derived payload attributes](#payload-attributes-derivation)
-(`payloadAttributes` argument) as well as the previous L2 state (`forkchoiceState` argument), defined by
-the [`engine_forkchoiceUpdatedV1`] method of the [Engine API]:
+## From Payload Attributes to L2 Block
 
-- `headBlockHash`: block hash of the last block of the L2 chain, according to the rollup driver.
-- `safeBlockHash`: same as `headBlockHash`.
-- `finalizedBlockHash`: the hash of the block whose number is `number(headBlockHash) - FINALIZATION_DELAY_BLOCKS` if
-  the number of that block is `>= L2_CHAIN_INCEPTION`, 0 otherwise (where `FINALIZATION_DELAY_BLOCKS == 50400`
-  (approximately 7 days worth of L1 blocks) and `L2_CHAIN_INCEPTION` is the [L2 chain inception] (the number of the
-  first L1 block for which an L2 block was produced). See the [Finalization Guarantees][finalization] section for more
-  details.
+Once the [payload attributes] for a given L1 block `B` have been built, and if we have already derived an L2 block from
+`B`'s parent block, then we can use the payload attributes to derive a new L2 block.
 
-[`engine_forkchoiceUpdatedV1`]: exec-engine.md#engine_forkchoiceUpdatedV1
+### Inductive Derivation Step
 
-Once this first API call completes, `engine_getPayloadV1` is used to fetch the full L2 block,
-as specified in the [Engine API].
+Let
 
-[`engine_getPayloadV1`]: exec-engine.md#engine_executepayloadv1
-[`ExecutionPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#executionpayloadv1
+- `refL2` be the (hash of) the current L2 chain head
+- `refL1` be the (hash of) the L1 block from which `refL2` was derived
+- `payloadAttributes` be some previously derived [payload attributes] for the L1 block with number `l1Number(refL1) + 1`
+
+Then we can apply the following pseudocode logic to update the state of both the rollup driver and execution engine:
+
+```pseudocode
```

Comment:
This is fine with me, but there's no syntax highlighting so it may be difficult to read. If Go is too verbose for snippets in the spec, we can try python.

Reply:
Agree with Python, it seems to be more or less the standard for pseudocode in ethereum

Reply:
I'll try using `javascript` since we use JSON notation at the top, I think it should highlight well then.

Reply:
Wrote my answer before seeing maurelian's. Feel free to suggest a modification to the syntax if you feel like it's warranted. I personally don't really see the need to change anything here.